### PR TITLE
Fix cpp ci

### PR DESCRIPTION
**Pull request by mreso**
## Description

Please read our [CONTRIBUTING.md](https://github.com/pytorch/serve/blob/master/CONTRIBUTING.md) prior to creating your first pull request.

Please include a summary of the feature or issue being fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the Unit or Integration tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
Logs for Test A

- [ ] Test B
Logs for Test B

## Checklist:

- [ ] Did you have fun?
- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: .github/workflows/ci-cpu-cpp.yml [hunk 1]
```diff
@@ -16,17 +16,42 @@ jobs:
...
+        with:
+          submodules: recursive
+      # - name: Install libtorch - macOS
+      #   if: matrix.os == 'macOS-latest'
+      #   run: |
+      #       brew install libtorch
       - name: Install dependencies
         run: |
-          python ts_scripts/install_dependencies.py --environment=dev --cpp
+          sudo apt update && python ts_scripts/install_dependencies.py --environment=dev --cpp
```
Comment (lxning) [inline .github/workflows/ci-cpu-cpp.yml:47]: can we check cmake version before we proceed?

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -31,6 +27,21 @@ find_package(fmt REQUIRED)
...
+  GIT_TAG 0.8.0 # Can be a tag (yaml-cpp-x.x.x), a commit hash, or a branch name (master)
+)
+FetchContent_GetProperties(yaml-cpp)
+
+if(NOT yaml-cpp_POPULATED)
+  message(STATUS "Fetching yaml-cpp...")
+  FetchContent_Populate(yaml-cpp)
+  add_subdirectory(${yaml-cpp_SOURCE_DIR} ${yaml-cpp_BINARY_DIR})
+endif()
+
```
Comment (lxning) [inline cpp/CMakeLists.txt:44]: we need build yaml-cpp with -DCMAKE_CXX_FLAGS="-fPIC" (see https://github.com/pytorch/serve/blob/master/cpp/build.sh#L133C84-L133C109)

File: cpp/README.md [hunk 3]
```diff
@@ -5,36 +5,34 @@
...
 mkdir model_store
 torchserve --ncs --start --model-store model_store
 ```
+
+### Clean the build directory
+To clean the build directory in order to rebuild from scratch simply delete the cpp/_build directory with
+```
+rm -rf cpp/_build
+```
+
```
Comment (lxning) [inline cpp/README.md:35]: can we keep this so that cx can clean up env in case of debugging?

File: cpp/build.sh [hunk 4]
```diff
@@ -115,58 +93,22 @@ function install_libtorch() {
...
-
-    if [ "$PLATFORM" = "Linux" ]; then
-      SUDO="sudo"
-    elif [ "$PLATFORM" = "Mac" ]; then
-      SUDO=""
-    fi
-
-    mkdir $YAML_CPP_BUILD_DIR
-    cd $YAML_CPP_BUILD_DIR
-    cmake $YAML_CPP_SRC_DIR -DYAML_BUILD_SHARED_LIBS=ON -DYAML_CPP_BUILD_TESTS=OFF -DCMAKE_CXX_FLAGS="-fPIC"
```
Comment (lxning) [inline cpp/build.sh:133]: We need build yaml-cpp with DCMAKE_CXX_FLAGS="-fPIC"

File: examples/cpp/llamacpp/CMakeLists.txt [hunk 5]
```diff
@@ -1,20 +1,20 @@
...
-    ${LLAMACPP_SRC_DIR}/ggml-alloc.o
-    ${LLAMACPP_SRC_DIR}/grammar-parser.o
-    ${LLAMACPP_SRC_DIR}/console.o
-    ${LLAMACPP_SRC_DIR}/build-info.o
-    ${LLAMACPP_SRC_DIR}/ggml-backend.o
-
+FetchContent_Declare(
+  llama.cpp
+  GIT_REPOSITORY https://github.com/ggerganov/llama.cpp
+  GIT_TAG b2241
```
Comment (lxning) [inline examples/cpp/llamacpp/CMakeLists.txt:9]: do we need build llama cpp with -DCMAKE_CXX_FLAGS="-fPIC"?

File: cpp/CMakeLists.txt (hunk 2 above)
Comment (mreso) [reply cpp/CMakeLists.txt:44]: @lxning we globally set the flag here https://github.com/pytorch/serve/blob/9db8ec6000d4e585348cdae4b459e7ff3e24eb83/cpp/CMakeLists.txt#L7 so it should be activated.

File: .github/workflows/ci-cpu-cpp.yml (hunk 1 above)
Comment (mreso) [reply .github/workflows/ci-cpu-cpp.yml:47]: @lxning not sure what you mean? Inside the workflow file? What would be the consequences of the check?

File: examples/cpp/llamacpp/CMakeLists.txt (hunk 5 above)
Comment (mreso) [reply examples/cpp/llamacpp/CMakeLists.txt:9]: See above

File: cpp/build.sh (hunk 4 above)
Comment (mreso) [reply cpp/build.sh:133]: See above

File: cpp/README.md (hunk 3 above)
Comment (mreso) [reply cpp/README.md:35]: @lxning Not sure if I understand. This is meant to be executed if you want to rebuild from scratch. Like a `make clean`. Can make this more clear in the description.

File: .github/workflows/ci-cpu-cpp.yml (hunk 1 above)
Comment (lxning) [reply .github/workflows/ci-cpu-cpp.yml:47]: install_dependency with option "cpp" will install cmake (eg. [mac](https://github.com/pytorch/serve/blob/master/ts_scripts/install_dependencies.py#L56), [linux](https://github.com/pytorch/serve/blob/master/ts_scripts/install_dependencies.py#L17)). Currently some cmake version can cause cpp compile error. Better to check cmake version before cijob proceed to save debug time.

File: cpp/README.md (hunk 3 above)
Comment (lxning) [reply cpp/README.md:35]: We haven't set clean rule at this moment. We need let cx know how to clean.

**Review by lxning — APPROVED** [review]

**Pull request merged**